File: wk6/java/evaluateReversePolishNotation.java
// problem link: https://leetcode.com/problems/evaluate-reverse-polish-notation/description/

class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> s = new Stack<>();
        for(String token : tokens) {
            if(token.equals("+"))
                s.push(s.pop() + s.pop());
            else if(token.equals("-")) {
                int a = s.pop(), b = s.pop();
                s.push(b - a);
            }
            else if(token.equals("*"))
                s.push(s.pop() * s.pop());
            else if(token.equals("/")) {
                int a = s.pop(), b = s.pop();
                s.push(b / a);
            }
            else
                s.push(Integer.parseInt(token));
        }
        return s.pop();
    }
}

File: wk6/cpp/evaluateReversePolishNotation.cpp
// problem link: https://leetcode.com/problems/evaluate-reverse-polish-notation/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        vector<int> v;
        for(auto const &t : tokens) {
            if(t == "+") {
                v[v.size() - 2] = v[v.size() - 2] + v[v.size() - 1];
                v.pop_back();
            }
            else if(t == "-") {
                v[v.size() - 2] = v[v.size() - 2] - v[v.size() - 1];
                v.pop_back();
            }
            else if(t == "*") {
                v[v.size() - 2] = v[v.size() - 2] * v[v.size() - 1];
                v.pop_back();
            }
            else if(t == "/") {
                v[v.size() - 2] = v[v.size() - 2] / v[v.size() - 1];
                v.pop_back();
            }
            else
                v.push_back(stoi(t));
        }
        return v[0];
    }
};